dataflow/flex-templates/kafka_to_bigquery/README.md

```diff
@@ -0,0 +1,337 @@
+
+Samples showing how to create and run an
+[Apache Beam](https://beam.apache.org/) template with a custom Docker image on
+[Google Cloud Dataflow](https://cloud.google.com/dataflow/docs/).
+
+## Before you begin
+
+If you are not familiar with Dataflow Flex templates, please see the
+[Streaming Beam SQL](../streaming-beam-sql/) sample first.
+
+Follow the
+[Getting started with Google Cloud Dataflow](../README.md)
+page, and make sure you have a Google Cloud project with billing enabled
+and a *service account JSON key* set up in your `GOOGLE_APPLICATION_CREDENTIALS`
+environment variable.
+Additionally, for this sample you need the following:
+
+1. Create a [Cloud Storage bucket](https://cloud.google.com/storage/docs/creating-buckets).
+
+    ```sh
+    export BUCKET="your-gcs-bucket"
+    gsutil mb gs://$BUCKET
+    ```
+
+1. Create a [BigQuery dataset](https://cloud.google.com/bigquery/docs/datasets).
+
+    ```sh
+    export PROJECT="$(gcloud config get-value project)"
+    export DATASET="beam_samples"
+    export TABLE="kafka_to_bigquery"
+
+    bq mk --dataset "$PROJECT:$DATASET"
+    ```
+
+1. Select the compute region and zone to use.
+
+    ```sh
+    # Select your default compute/region, or default to "us-central1".
+    export REGION=${"$(gcloud config get-value compute/region)":-"us-central1"}
+
+    # Select your default compute/zone, or default to "$REGION-a".
+    # Note that the zone *must* be in $REGION.
+    export ZONE=${"$(gcloud config get-value compute/zone)":-"$REGION-a"}
+    ```
+
+1. Clone the `java-docs-samples` repository.
+
+    ```sh
+    git clone https://github.com/GoogleCloudPlatform/java-docs-samples.git
+    ```
+
+1. Navigate to the sample code directory.
+
+    ```sh
+    cd java-docs-samples/dataflow/flex-templates/kafka_to_bigquery
+    ```
+
+## Kafka to BigQuery sample
+
+This sample shows how to deploy an Apache Beam streaming pipeline that reads
+[JSON encoded](https://www.w3schools.com/whatis/whatis_json.asp) messages from
+[Apache Kafka](https://kafka.apache.org/), decodes them, and writes them into a
+[BigQuery](https://cloud.google.com/bigquery) table.
+
+For this, we need two parts running:
+
+1. A Kafka server container accessible through an external IP address.
+   This services publishes messages to a topic.
+
+    * [kafka/Dockerfile](kafka/Dockerfile)
+    * [kafka/start-kafka.sh](kafka/start-kafka.sh)
+    * [kafka/create-topic.sh](kafka/create-topic.sh)
+
+2. An Apache Beam streaming pipeline running in Dataflow Flex Templates.
+   This subscribes to a Kafka topic, consumes the messages that are published
+   to that topic, processes them, and writes them into a BigQuery table.
+
+    * [Dockerfile](Dockerfile)
+    * [KafkaToBigQuery.java](src/main/java/org/apache/beam/samples/KafkaToBigQuery.java)
+    * [pom.xml](pom.xml)
+    * [metadata.json](metadata.json)
+
+### Starting the Kafka server
```

Comment:
Wow that's a lot of steps, I wonder if it wouldn't be easier on users to just use a script that asks them what it needs to know?

Reply:
Decided not to script this in the end. Find explanation below.